### ISSUE TITLE
Combining Box, Point, and Text Prompts to Improve Segmentation Mask Accuracy

### ISSUE THREAD

**Issue by ahmoub** [facebookresearch/sam3#395]
Hello,

I have a question regarding the use of prompts. I first used a **bounding box** as a prompt to detect a part in an image. Afterwards, I used **only point prompts**. My question is the following: **is it possible to combine point and box prompts to improve the accuracy of the generated masks?**

I know that it is possible to use a bounding box and add points that are treated as **negative prompts**, but is it also possible to use a bounding box and add **positive point prompts** in order to further refine and improve the segmentation results?

I have the same question regarding **text prompts**: is it possible to combine a text prompt with other types of prompts (such as box or point prompts) to improve the accuracy of the resulting masks?

**Comment by heyoeyo** [comment]
There are 2 different ways that the SAMv3 model can do point/box segmentation. There's the 'old' way that matches the v1 & v2 models and the newer 'detector' way (which is what happens when using text). Both support combining prompts, though the newer detector approach (afaik) doesn't have an easy way to provide points as inputs (though it is supported).

> I know that it is possible to use a bounding box and add points that are treated as negative prompts, but is it also possible to use a bounding box and add positive point prompts in order to further refine and improve the segmentation results?

The 'old' way can do this pretty easily. There's an example in one of the notebooks:
https://github.com/facebookresearch/sam3/blob/main/examples/sam3_for_sam1_task_example.ipynb
Under the section **Combining points and boxes** (about 2/3rds of the way down the notebook). That example uses a negative point label (e.g. `input_label = np.array([0])`), but changing it to a `1` makes it a positive point.

> I have the same question regarding text prompts: is it possible to combine a text prompt with other types of prompts (such as box or point prompts) to improve the accuracy of the resulting masks?

It should be possible to combine text and boxes by adding them both to the 'inference state':
```python
state = processor.set_image(image)
state = processor.set_text_prompt(state=state, prompt=...)
state = processor.add_geometric_prompt(state=state, box=...)
```
The model does support points internally, but there doesn't seem to be a convenient 'add points' function, so you'd have to do that manually. If you did want to implement this, it's probably easiest to copy the existing [add_geometric_prompt](https://github.com/facebookresearch/sam3/blob/5eb25fb54b70ec0cb16f949289053be091e16705/sam3/model/sam3_image_processor.py#L128) code, and then just swap out the [append_boxes](https://github.com/facebookresearch/sam3/blob/5eb25fb54b70ec0cb16f949289053be091e16705/sam3/model/sam3_image_processor.py#L150) part for [append_points](https://github.com/facebookresearch/sam3/blob/5eb25fb54b70ec0cb16f949289053be091e16705/sam3/model/geometry_encoders.py#L354).

**Comment by JRGit4UE** [comment]
Has anyone tried to combine more than one box along with points?
I use the https://github.com/facebookresearch/sam3/blob/main/examples/sam3_for_sam1_task_example.ipynb as a starter to combine boxes and points, but fail to extend the example to use **more than 1 box / points**? 
``` python
input_boxes = np.array([
    [188, 148, 500, 703],
    [650, 148, 1100, 705],
])
input_point = np.array([[0, 0], [1090, 535]])
input_label = np.array([0,1])

masks, scores, _ = model.predict_inst(
    inference_state,
    point_coords=input_point, # None,
    point_labels=input_label,  # None,
    box=input_boxes[None, :],
    multimask_output=False,
)
``` 
or other tensor combinations of boxes and points fail with dimension mismatch errors..

**Comment by heyoeyo** [comment]
> Has anyone tried to combine more than one box along with points?

As-is, the code doesn't support multiple boxes as a single prompt. The model is capable of it, but it requires modifying the code. Though in practice, using multiple boxes doesn't seem to work well (imo).

See the same issue from SAMv2:
https://github.com/facebookresearch/segment-anything/issues/620

That being said, you can use multiple points with a single box (which can work well), or you can have multiple boxes if they're meant to be separate prompts (i.e. generate independent masks). The shapes have to be:
Boxes: Bx4
Points: BxNx2
Labels: BxN

Where B is the 'batch' size (i.e. the number of independent masks), N is the number of points.
So for the example you gave, changing the prediction call to some thing like this should work:
```python
masks, scores, _ = model.predict_inst(
    inference_state,
    point_coords=input_point[:, None, :],
    point_labels=input_label[:, None],
    box=input_boxes,
    multimask_output=False,
)
```
Though this would be generating two masks, each using a single box and single point (maybe not what you want).

**Comment by syhwang1231** [comment]
@ahmoub Did you succeed at combining a text prompt and a point prompt? I'm trying the same thing but in reverse order; point prompt - then - text prompt.

I implemented `add_point_prompt` method, following this issue and issue #337, like this:
```python
    @torch.inference_mode()
    def add_point_prompt(self, point, label: bool, state: Dict):
        '''
        No explicit method for adding a point prompt, so implemented manually by referring to issue #337, 395

        Args:
            point:
            label (bool): pos / neg 
            state (dict): inference_state
        '''
        if "backbone_out" not in state:
            raise ValueError("You must call set_image before set_text_prompt")
        
        if "language_features" not in state["backbone_out"]:
            dummy_text_outputs = self.model.backbone.forward_text(
                ["visual"], device=self.device
            )
            state["backbone_out"].update(dummy_text_outputs)
        
        if "geometric_prompt" not in state:
            state["geometric_prompt"] = self.model._get_dummy_prompt()

        pts_tensor = torch.tensor(point, device=self.device, dtype=torch.float32).view(1, 1, 2)
        labels_tensor = torch.tensor([label], device=self.device, dtype=torch.bool).view(1, 1)
        state["geometric_prompt"].append_points(pts_tensor, labels_tensor)
        return self._forward_grounding(state)
```

And in my main code, I call the methods in this order:

```python
inference_state = processor.set_image(image)
inference_state = processor.add_point_prompt(point=input_point, label=True, state=inference_state)
inference_state = processor.set_text_prompt(state=inference_state, prompt=text_prompt) 
```

But when I check my result, something is not working right. It seems like the point prompt works as a negative point when it should work as a positive point prompt.

<img width="3285" height="1176" alt="Image" src="https://github.com/user-attachments/assets/b2215b3f-ccaf-4efc-9599-318741b48b7c" />

<img width="3285" height="1176" alt="Image" src="https://github.com/user-attachments/assets/f4bd68d2-bb2a-4243-9d81-40cef8013c4c" />
I'm thinking something is wrong with the `add_point_prompt()` method, but I just can't find out what's making it work like this.

**Comment by heyoeyo** [comment]
> It seems like the point prompt works as a negative point when it should work as a positive point prompt

It might be that there's a discrepancy between where the point is drawn and what's being given to the model (i.e. the model is evaluating a point that isn't located where the green dot is being drawn in the images). That might happen if the point is given in pixel units, for example.

As for combining text & points, I think it only tends to work if the point prompt is helping to refine a text prompt that is already being specific. For example in the repo [test image](https://github.com/facebookresearch/sam3/blob/5eb25fb54b70ec0cb16f949289053be091e16705/assets/images/test_image.jpg), if you use a prompt of 'blue shirt' it doesn't work. If you use only a point prompt on a blue shirt, it segments all of the shirts. But if you combine the text and point prompt, it correctly segments only the blue shirts.
For the examples given above where the text is very general, then I think point prompts won't (typically) work to select a specific object. It may be better to use the [SAMv1 point prompts](https://github.com/facebookresearch/sam3/blob/5eb25fb54b70ec0cb16f949289053be091e16705/examples/sam3_for_sam1_task_example.ipynb), which is designed for this sort of task.
